text/0000-variadic-generics.md

```diff
@@ -0,0 +1,444 @@
+let (_, ...t1) = t0 // What is t1? is it (&&u16, &&u8)?
+```
+
+This will increase the number of instructions need to unpack, as a
+new `Tuple` will need to be constructed.
+
+#### Solution 2
+[solution-2]: #solution-2
+
+In theory, instead of merely capturing the tail by reference we could
+construct a new `Tuple` and move each value of the tail to the new
+structure with the added padding.
+
+```rust
+let some_tuple = (42u8, 0u8, 65535u16, 0u8);
+let (_, ...t0) = some_tuple // t0: (u8, u16, u8)
+let (_, ...t1) = t0 // t1: (u16, u8)
+
+do_something(some_tuple); // ERROR: Use of moved value: `some_tuple`
+```
+
+**Limitations**: This will increase the number of instructions need to
+unpack, as a new `Tuple` will need to be constructed.
+
+#### Solution 3
+[solution-3]: #solution-3
+
+The unpacked tuple has some "special sauce" and is not actually allocated.
+A new `TyKind` is created and the compiler essentially transforms the
+following:
+
+```rust
+let some_tuple = (42u8, 0u8, 65535u16, u8);
+let (_, ...tail) = some_tuple;
+
+do_something(...tail);
+```
+
+Into something like the following:
+
+```rust
+let some_tuple = (42u8, 0u8, 65535u16, u8);
+let _ = some_tuple.0
+
+do_something(some_tuple.1, some_tuple.2, some_tuple.3);
+```
+
+Any attempt to use `tail` as a "real" `Tuple` results in a compiler
+error.
+
+**Limitations**: This feels very similar to C\+\+11's Parameter Packs.
+We're creating a new type with its own behavior and missing out on the
+chance to improve the ergonomics of `Tuple`.
+
+#### Solution Discussion
+
+ - Both solutions that allow unpacking a `Tuple` are `O(n)`. Again, we
+   unfortunately can not use a simple `memcpy` due to the possibility
+   of padding changes.
+ - If Solution 1 or 2 is chosen, and the `O(n)` operation of unpacking the
+   original `Tuple` and constructing the tail `Tuple` does not get optimized
+   out by LLVM/MIR when the unpacked tail `Tuple` is not used, a new
+   optimization pass should be written. For example, if the tail `Tuple` is
+   immediately passed to a function, there may be no need to construct
+   the `Tuple` for the `tail`.
+
+#### Open Questions
+
+How do we avoid generating a massive number of functions? For example,
+given the simple variadic generic function:
+
+```rust
+pub fn add<...T>(x: u8, ...args: T) -> u8 {
+    x + add(...args)
+}
+```
+
+And the use
+
+```rust
+add(1, 2, 3, 4, 5);
+```
+
+We need to avoid generating a set of recursive functions like the
+following:
+
+```rust
+fn add_2(x: (u8, u8)) -> u8 {
+    (x.0 + x.1)
+}
+
+fn add_1(x: (u8, u8, u8)) -> u8 {
+    x.0 + add_2((x.1, x.2))
+}
+
+fn add_0(x: (u8, u8, u8, u8)) -> u8 {
+    x.0 + add_1((x.1, x.2, x.3))
+}
+
+pub fn add(x: (u8, u8, u8, u8, u8)) -> u8 {
+    x.0 + add_0((x.1, x.2, x.3, x.4))
+}
+```
+
+**TODO(dlrobertson)**: I do not see a way to avoid this if we stick to
+the strict `Tuple` implementation of variadic generics. Again, [solution-3]
+seems like the best choice.
```

Comment:
How does solution-3 avoid the monomorphization problems? In this case, inlining will most likely get rid of all the function calls and strip things down to just a constant folded value `15u8`.

Reply:
I did some more research today and I don't think this is a strike against tuple solutions as long as the optimization pass that avoids unnecessary tuple construction/destruction is implemented. E.g. generating the following code would be acceptable (it is what clang does).

```rust
fn add_2(a0: u8, a1: u8) -> u8 {
    a0 + a1
}
fn add_1(a0: u8, a1: u8, a2: u8) -> u8 {
    a0 + add_2(a1, a2)
}
fn add_0(a0: u8, a1: u8, a2: u8, a3: u8) -> u8 {
    a0 + add_1(a1, a2, a3)
}
pub fn add(a0: u8, a1: u8, a2: u8, a3: u8, a4: u8) -> u8 {
    a0 + add_0(a1, a2, a3, a4)
}
```